src/template-shadowroot.ts

```diff
@@ -0,0 +1,119 @@
+/**
+ * @license
+ * Copyright (c) 2019 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+let hasNative: boolean|undefined;
+export function hasNativeDeclarativeShadowRoots(): boolean {
+  if (hasNative === undefined) {
+    const div = document.createElement('div');
+    div.innerHTML = `<div><template shadowroot="open"></template></div>`;
+    hasNative = !!div.firstElementChild!.shadowRoot;
+  }
+  return hasNative;
+}
+
+/*
+ * Traverses the DOM to find all <template> elements with a `shadowroot`
+ * attribute and move their content into a ShadowRoot on their parent element.
+ *
+ * This processing is done depth-first so that when top-level <template>
```

Comment:
Do you know if the depth first processing is actually part of the spec? It seems like that would be bad for performance (aka streaming).

I suspect manually walking the DOM like this is significantly slower than recursively querySelector'ing for templates.

Reply:
The processing happens upon reaching each `</template>` so in effect it does happen bottom up in the (current proposed) spec. This works fine for streaming.

querySelectorAll is totally something we should put into the comparison benchmarks. added a note.

Reply:
The problem with query selector is that you have to walk the nested templates anyway. Worth a try, but it's not like you can just get all the templates in one call.

Reply:
You can QSA the contents of each template and having the nested content separate should help with the ordering. I'd be surprised if this wasn't faster, but that's why we have tachometer =)